Implement switchable time format for decks

poelzi opened the pull request:
The display format on the decks can now be configured.
Implement first alternative display based on kiloseconds.

If you mixx tracks based on first and last downbeat, it is unnecessary complicated to use minutes. The first alternative display shows "k.sss:zzz" following roughly the idea of the kilosecond so it can be much easier calculated when the next track should be started.

poelzi committed:
Implement switchable duration format

The display format on the decks can now be configured.
Implement alternative display based on kiloseconds.

daschuer commented:
Thank you, nice Idea. Can you file a bug for this, than we can track this feature for a change log.

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -495,11 +513,31 @@ void DlgPrefDeck::slotRateRampingModeLinearButton(bool checked) {
+    m_pConfig->set(ConfigKey("[Controls]","TimeFormat"), ConfigValue(v));
+    comboBoxTimeFormat->setCurrentIndex(
+                comboBoxTimeFormat->findData(i));
+
```
extra new line

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -495,11 +513,31 @@ void DlgPrefDeck::slotRateRampingModeLinearButton(bool checked) {
+
+void DlgPrefDeck::slotTimeFormatChanged(int index) {
+    int mode = comboBoxTimeFormat->itemData(index).toInt();
+    m_timeDisplayFormat = static_cast<TrackTime::DisplayFormat>(mode);
```
I think you can omit the m_timeDisplayFormat here and read the comboBoxTimeFormat directly

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -495,11 +513,31 @@ void DlgPrefDeck::slotRateRampingModeLinearButton(bool checked) {
+}
+
 void DlgPrefDeck::slotApply() {
     double timeDisplay = static_cast<double>(m_timeDisplayMode);
```
I think we can get also rid of m_timeDisplayMode

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -495,11 +513,31 @@ void DlgPrefDeck::slotRateRampingModeLinearButton(bool checked) {
+
+}
+
+void DlgPrefDeck::slotTimeFormatChanged(int index) {
```
I think we should get rid of this. It is anyway confusing to have this function twice as int and double overload.

daschuer commented on src/preferences/dialog/dlgprefdeck.h:
```diff
@@ -22,6 +22,10 @@ namespace TrackTime {
         ElapsedAndRemaining,
     };
+    enum class DisplayFormat {
+        DEFAULT,
```
DEFAULT is meaningless here. How about WALL_CLOCK or HOUR_MINUTES_SECONDS or TRADITIONAL

daschuer commented on src/util/duration.cpp:
```diff
@@ -48,5 +48,36 @@ QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+
+    QString durationString =
+            (days > 0 ? (QString::number(days) %
+                         QLatin1String("d, ")) : QString()) %
```
Why do we need to deal with days? Can't we just count seconds?

daschuer replied:
I think we schould remove the day part entirely. No track will have this length, and if there is what we have unlikely the screen space for this.
I personally would prefer to see 
30:10:01,86 instead of 1d,6:10:01,86

daschuer commented on src/util/duration.cpp:
```diff
@@ -48,5 +48,36 @@ QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+    QString durationString =
+            (days > 0 ? (QString::number(days) %
+                         QLatin1String("d, ")) : QString()) %
+            QString("%1.%2").arg(kilos, 0, 10).arg(seconds, 3, 'f', 0, QLatin1Char('0'));
```
QLocale::toString() can already do this in the correct format.

daschuer commented on src/preferences/dialog/dlgprefdeck.cpp:
```diff
@@ -106,6 +106,24 @@ DlgPrefDeck::DlgPrefDeck(QWidget * parent, MixxxMainWindow * mixxx,
+    comboBoxTimeFormat->clear();
+    comboBoxTimeFormat->addItem(tr("hh:mm:ss.zzz"),
+                                static_cast<int>(TrackTime::DisplayFormat::DEFAULT));
+    comboBoxTimeFormat->addItem(tr("k.sss:zz"),
```
Or plain ks like use k.ssszz Or "k.sss zz" with space as group separator as defined in the SI system 

https://en.wikipedia.org/wiki/Decimal_separator#Digit_grouping

daschuer replied:
Mmm I have just checked, currently I have the 00:00.00 format even in German. Now I can remember that we have discussed it before to use 00:00,00 for us, but this looks unusual, because all hardware devices have the dot, even though it would be correct. 

When reading your PR I was stumbeling over "k.sss:zzz", just because is somehow ambiguous in this case.

7.332:84 Colon 
7,332 84 Space
7,332 84 Punctuation Space
7,332 84 Thin Space
7,332 84 Hair Space

I think I prefer for now the localized thin Space version as recommended by SI-System.

poelzi committed:
Cleanup and Hectoseconds

Fix comments regarding style.
Also add hectosecond formatting. For mixing, 100 seconds is actually a quite useful
seperator.

daschuer commented on src/util/duration.cpp:
```diff
@@ -17,6 +17,9 @@ static const qint64 kSecondsPerDay = 24 * kSecondsPerHour;
 } // namespace
 
+// static
+QString DurationBase::kCentisecondSeperator = QString::fromUtf8("\u2009");
```
This requires a comment. 
// Unicode for THIN SPACE

I am also a bit confused why it works. The UTF8 character is 0xe28089 
"\u2009" should be already UTF-16, the internal codec of QString. Does the compiler turns it into utf8 and than fromUtf8 back at rune time? 
Did you try QChar(0x2009); ?

daschuer commented:
It works nice. Should the library column also follow the settings? 
I cannot really deal with the hectosecond format. The dot for 100 s is useless for me. It is no problem at all, because it is just an option, but why not replace it by, or add a plain seconds format?

daschuer commented:
A Beats time format or better bars, if we have it one day would also be a nice benefit.
I have never worked with that, but I can't count seconds if a 130 bpm beat is hammering in the same time.

poelzi committed:
add simple second display.

rename formatSeconds into formatTime to reflect function better.

poelzi committed:
fixup tests. Try some unrelated character as seperator in hectoseconds.

poelzi committed:
Merge remote-tracking branch 'origin/master' into feature/seconds

Be-ing commented on src/util/duration.cpp:
```diff
@@ -32,12 +36,85 @@ QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+}
+
+QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+    if (dSeconds < 0.0) {
```
Use VERIFY_OR_DEBUG_ASSERT macro

Be-ing commented on src/util/duration.cpp:
```diff
@@ -32,12 +36,85 @@ QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+
+// static
+QString DurationBase::formatKiloSeconds(double dSeconds, Precision precision) {
+    if (dSeconds < 0.0) {
```
use VERIFY_OR_DEBUG_ASSERT macro

Be-ing commented on src/util/duration.cpp:
```diff
@@ -32,12 +36,85 @@ QString DurationBase::formatSeconds(double dSeconds, Precision precision) {
+
+// static
+QString DurationBase::formatHectoSeconds(double dSeconds, Precision precision) {
+    if (dSeconds < 0.0) {
```
use VERIFY_OR_DEBUG_ASSERT macro

Be-ing commented on src/util/duration.cpp:
```diff
@@ -18,7 +18,11 @@ static const qint64 kSecondsPerDay = 24 * kSecondsPerHour;
+
+// static
+QString DurationBase::formatTime(double dSeconds, Precision precision) {
     if (dSeconds < 0.0) {
```
use VERIFY_OR_DEBUG_ASSERT macro

beenisss replied:
@Be-ing when I replace these I get `Critical [Main]: DEBUG ASSERT: "dSeconds < 0.0" in function static QString mixxx::DurationBase::formatTime(double, mixxx::DurationBase::Precision) at src/util/duration.cpp:26` when I attempt to run. Any ideas?

Be-ing reviewed (commented):
Thanks for working on this. LGTM other than a few places where the VERFIY_OR_DEBUG_ASSERT macro could be used. @daschuer any last comments?

Be-ing commented:
Ping @poelzi. Could you make the small edits mentioned above? I'd like to merge this by Saturday for 2.2.

beenisss commented:
@poelzi would you be happy for somebody else to pick this up and finish it off?

Be-ing commented:
Printing with `qDebug` is redundant with the VERIFY_OR_DEBUG_ASSERT macro. That macro will print a warning to the log if the condition is false. In these cases, yes, simply replacing the `if` statements with `VERIFY_OR_DEBUG_ASSERT` is sufficient because the `if` statements already handle what to do in case of an error.

beenisss commented:
I can't get this to build. I get a dozen or so errors, none of which seem related (at least not directly) to the files that were changed:
```
src/library/analysisfeature.cpp:72:5: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:74:5: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:76:5: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:78:5: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:81:5: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:84:5: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:86:5: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:142:9: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:146:9: error: no matching member function for call to 'connect'
src/library/analysisfeature.cpp:95:54: error: cannot initialize a parameter of type 'QWidget *' with an lvalue of type 'DlgAnalysis *'
src/library/dlganalysis.h:64:18: error: field has incomplete type 'QButtonGroup' (x3)
```

Any idea why that would happen? I completely cleaned beforehand including running `rm -R config.log .sconsign.dblite .sconf_temp .sconsign.branch`

beenisss commented on src/util/duration.cpp:
```diff
@@ -3,7 +3,7 @@
 #include <QStringBuilder>
 #include <QTime>
-
+#include <math.h>
```
I'm a bit confused by this addition. It has no path and the syntax doesn't match the other entries. Besides which, two lines below is the statement `#include "util/math.h"` - is that a different file?

rryan replied:
`"util/math.h"` (in mixxx's codebase) is preferable to `<math.h>` (the [C standard library header](http://www.cplusplus.com/reference/cmath/)), since it includes Mixxx-specific helpers / shims.

rryan commented:
> Here you go, I think:

Those aren't the same errors you were seeing before?

beenisss commented:
I'm not sure what you mean, I think I've probably misunderstood what you asked for or have misled you with what I originally posted.

I'm currently working on reimplimenting this step by step and building as I go. I've noticed a couple of weird things in the updated code in the process of doing so, maybe once those are ironed out it'll work. I'm happy to post up anything that's useful for troubleshooting in the meantime though.

beenisss commented:
Well, I've applied the updates from this PR to a new branch manually with some changes that were minor and almost entirely cosmetic, and somehow it builds without issue.
 
¯\\\_(ツ)_/¯

ronso0 commented:
In my personal branch I've changed `mixxx::Duration::Precision::CENTISECONDS` to `...::SECONDS` in /src/widget/wnumberpos.cpp to get rid of the rather distracting centiseconds.

Can we add another format `hh:mm:ss` and maybe call it "Traditional, coarse"?

benis commented:
@ronso0 I couldn't get this to build so I reimplemented it and have created a new PR at #1915 where I've copied your request.

There are a couple of things still need ironing out, feel free to get other people involved.

benis commented:
In light of the above I think this can be closed.

poelzi commented:
When I developed this, I tried different characters. I think I spent an hour trying different combinations for this not really usually formats. I chose those because after giving each version a try of 10 minutes or so, I found those ok. I'm quite open to some better ideas :)

daschuer commented:
I can build this branch merged to master. @benis What is your issue? 

Appveyor is also able to build this. However some tests fails: 
[  FAILED  ] DurationUtilTest.FormatKiloSeconds
[  FAILED  ] DurationUtilTest.FormatHectoSeconds

Travis has timed out.

daschuer commented:
closed in favor of https://github.com/mixxxdj/mixxx/pull/1985
